src/Microsoft.CodeQuality.Analyzers/Core/Security/MarkVerbHandlersWithValidateAntiforgeryTokenAnalyzer.cs

```diff
@@ -0,0 +1,360 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using System;
+
+namespace Microsoft.CodeQuality.Analyzers.Security
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class MarkVerbHandlersWithValidateAntiforgeryTokenAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA3147";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(
+            nameof(MicrosoftSecurityAnalyzersResources.MarkVerbHandlersWithValidateAntiforgeryTokenTitle),
+            MicrosoftSecurityAnalyzersResources.ResourceManager,
+            typeof(MicrosoftSecurityAnalyzersResources));
+
+        private static readonly LocalizableString NoVerbsMessage = new LocalizableResourceString(
+            nameof(MicrosoftSecurityAnalyzersResources.MarkVerbHandlersWithValidateAntiforgeryTokenNoVerbsMessage),
+            MicrosoftSecurityAnalyzersResources.ResourceManager,
+            typeof(MicrosoftSecurityAnalyzersResources));
+
+        private static readonly LocalizableString NoVerbsNoTokenMessage = new LocalizableResourceString(
+            nameof(MicrosoftSecurityAnalyzersResources.MarkVerbHandlersWithValidateAntiforgeryTokenNoVerbsNoTokenMessage),
+            MicrosoftSecurityAnalyzersResources.ResourceManager,
+            typeof(MicrosoftSecurityAnalyzersResources));
+
+        private static readonly LocalizableString GetAndTokenMessage = new LocalizableResourceString(
+            nameof(MicrosoftSecurityAnalyzersResources.MarkVerbHandlersWithValidateAntiforgeryTokenGetAndTokenMessage),
+            MicrosoftSecurityAnalyzersResources.ResourceManager,
+            typeof(MicrosoftSecurityAnalyzersResources));
+
+        private static readonly LocalizableString GetAndOtherAndTokenMessage = new LocalizableResourceString(
+            nameof(MicrosoftSecurityAnalyzersResources.MarkVerbHandlersWithValidateAntiforgeryTokenGetAndOtherAndTokenMessage),
+            MicrosoftSecurityAnalyzersResources.ResourceManager,
+            typeof(MicrosoftSecurityAnalyzersResources));
+
+        private static readonly LocalizableString VerbsAndNoTokenMessage = new LocalizableResourceString(
+            nameof(MicrosoftSecurityAnalyzersResources.MarkVerbHandlersWithValidateAntiforgeryTokenVerbsAndNoTokenMessage),
+            MicrosoftSecurityAnalyzersResources.ResourceManager,
+            typeof(MicrosoftSecurityAnalyzersResources));
+
+        private static readonly DiagnosticDescriptor NoVerbsRule = new DiagnosticDescriptor(
```

Comment:
Consider adding the `description` and `helpUri` arguments to all diagnostic descriptor constructor invocations. Do we plan to add public documentation on docs.microsoft.com for this rule and all the cases for which we report this diagnostic?

Reply:
I suppose we should; I'll have to follow up with who's writing documentation.

Reply:
Tagging @gewarren, who is our documentation expert for analyzer spaces. @dotpaul Can you file an issue on https://github.com/MicrosoftDocs/visualstudio-docs/issues to track this work? You should sync up offline with Genevieve for the actual content and source examples.

Reply:
Sure, created https://github.com/MicrosoftDocs/visualstudio-docs/issues/1262